File: gpt-4-turbo/stage2/loop_construct_private_clause.c
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>

#ifndef SEED
#define SEED 12345
#endif

#ifndef NUM_TEST_CALLS
#define NUM_TEST_CALLS 10
#endif

int test1(){
    int err = 0;
    srand(SEED);
    int a[100];
    int initialValue = 5;

    // Initialize the array to check later if modifications have been isolated by private tmp
    for (int i = 0; i < 100; ++i) {
        a[i] = 0;
    }

    // Test starts
    #pragma acc data copy(a)
    {
        #pragma acc parallel loop private(initialValue)
        for (int i = 0; i < 100; ++i) {
            int tmp = initialValue; // tmp is private to each iteration
            tmp += i; // Modify tmp in a way that depends on the loop variable
            if (tmp != initialValue + i){
                err++; // This should never happen if tmp is truly private
            }
            a[i] = tmp; // Write the result back to check after
        }
    }

    // Verify that changes to tmp in the loop were isolated between iterations
    for (int i = 0; i < 100; ++i) {
        if (a[i] != initialValue + i) {
            err++;
        }
    }

    return err;
}

int main(){
    int failcode = 0;
    int failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed += test1();
    }
    if (failed != 0){
        failcode += (1 << 0);
    }
    return failcode;
}

File: gpt-4-turbo/stage2/loop_construct_private_clause.f90
! This program tests the proper implementation of the 'private' clause in OpenACC loop constructs
PROGRAM test_private_clause
  USE openacc
  IMPLICIT NONE
  INTEGER, PARAMETER :: n = 10
  REAL, DIMENSION(n,n) :: a, b, c
  REAL :: tmp, div
  INTEGER :: i, j, k
  INTEGER :: errors

  ! Initialize arrays and div
  div = 2.0
  a = 0.0
  b = 1.0
  c = 2.0
  errors = 0

  !$acc parallel num_gangs(1) copy(a) copyin(b,c) private(tmp)
  !$acc loop gang
  DO k = 1, n
    !$acc loop worker private(tmp)
    DO j = 1, n
      tmp = b(j,k) + c(j,k)
      !$acc loop vector
      DO i = 1, n
        a(i,j,k) = a(i,j,k) + tmp / div
      ENDDO
    ENDDO
  ENDDO
  !$acc end parallel
  
  ! Verify Results - Ideally, tmp should have been private and the computation reflects that
  DO k = 1, n
    DO j = 1, n
      DO i = 1, n
        IF (ABS(a(i,j,k) - ((b(j,k) + c(j,k)) / div)) > 1.0E-6) THEN
          errors = errors + 1
        END IF
      ENDDO
    ENDDO
  ENDDO

  IF (errors == 0) THEN
    PRINT *, "Test Passed."
  ELSE
    PRINT *, "Test Failed."
  END IF
END PROGRAM test_private_clause